File: provable-sdk-js/src/hash.test.ts
/**
 * Tests for hash module
 */

import { describe, it, expect } from 'vitest';
import { hash, keccak256, hash_str, keccak256_str, sha256, sha256_str } from './hash';

describe('hash', () => {
  describe('keccak256', () => {
    it('should hash empty data', () => {
      const data = new Uint8Array([]);
      const result = keccak256(data);
      expect(result).toBe('c5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470');
    });

    it('should hash simple data', () => {
      const data = new TextEncoder().encode('hello');
      const result = keccak256(data);
      expect(result).toHaveLength(64); // 32 bytes in hex
      expect(result).toMatch(/^[0-9a-f]{64}$/);
    });

    it('should produce consistent results', () => {
      const data = new TextEncoder().encode('test');
      const hash1 = keccak256(data);
      const hash2 = keccak256(data);
      expect(hash1).toBe(hash2);
    });

    it('should produce different hashes for different data', () => {
      const data1 = new TextEncoder().encode('hello');
      const data2 = new TextEncoder().encode('world');
      expect(keccak256(data1)).not.toBe(keccak256(data2));
    });
  });

  describe('keccak256_str', () => {
    it('should hash empty string', () => {
      const result = keccak256_str('');
      expect(result).toBe('c5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470');
    });

    it('should hash simple string', () => {
      const result = keccak256_str('hello');
      expect(result).toHaveLength(64);
      expect(result).toMatch(/^[0-9a-f]{64}$/);
    });

    it('should match keccak256 with encoded data', () => {
      const str = 'test string';
      const data = new TextEncoder().encode(str);
      expect(keccak256_str(str)).toBe(keccak256(data));
    });
  });

  describe('hash aliases', () => {
    it('hash should be same as keccak256', () => {
      const data = new TextEncoder().encode('test');
      expect(hash(data)).toBe(keccak256(data));
    });

    it('hash_str should be same as keccak256_str', () => {
      expect(hash_str('test')).toBe(keccak256_str('test'));
    });
  });

  describe('sha256', () => {
    it('should hash empty data', async () => {
      const data = new Uint8Array([]);
      const result = await sha256(data);
      expect(result).toBe('e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855');
    });

    it('should hash simple data', async () => {
      const data = new TextEncoder().encode('hello');
      const result = await sha256(data);
      expect(result).toHaveLength(64);
      expect(result).toMatch(/^[0-9a-f]{64}$/);
    });

    it('should produce consistent results', async () => {
      const data = new TextEncoder().encode('test');
      const hash1 = await sha256(data);
      const hash2 = await sha256(data);
      expect(hash1).toBe(hash2);
    });

    it('should produce different hashes for different data', async () => {
      const data1 = new TextEncoder().encode('hello');
      const data2 = new TextEncoder().encode('world');
      const hash1 = await sha256(data1);
      const hash2 = await sha256(data2);
      expect(hash1).not.toBe(hash2);
    });
  });

  describe('sha256_str', () => {
    it('should hash empty string', async () => {
      const result = await sha256_str('');
      expect(result).toBe('e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855');
    });

    it('should hash simple string', async () => {
      const result = await sha256_str('hello');
      expect(result).toHaveLength(64);
      expect(result).toMatch(/^[0-9a-f]{64}$/);
    });

    it('should match sha256 with encoded data', async () => {
      const str = 'test string';
      const data = new TextEncoder().encode(str);
      expect(await sha256_str(str)).toBe(await sha256(data));
    });
  });

  describe('hash algorithms comparison', () => {
    it('keccak256 and sha256 should produce different hashes', async () => {
      const data = new TextEncoder().encode('test');
      const keccakHash = keccak256(data);
      const sha256Hash = await sha256(data);
      expect(keccakHash).not.toBe(sha256Hash);
    });
  });
});

File: provable-sdk-go/hash_test.go
package provable

import (
	"bytes"
	"regexp"
	"testing"
)

func TestKeccak256(t *testing.T) {
	t.Run("hash empty data", func(t *testing.T) {
		data := []byte{}
		result := Keccak256(data)
		expected := "c5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470"
		if result != expected {
			t.Errorf("Keccak256() = %v, want %v", result, expected)
		}
	})

	t.Run("hash simple data", func(t *testing.T) {
		data := []byte("hello")
		result := Keccak256(data)
		if len(result) != 64 {
			t.Errorf("Keccak256() length = %v, want 64", len(result))
		}
		matched, _ := regexp.MatchString("^[0-9a-f]{64}$", result)
		if !matched {
			t.Error("Keccak256() result is not valid hex")
		}
	})

	t.Run("produce consistent results", func(t *testing.T) {
		data := []byte("test")
		hash1 := Keccak256(data)
		hash2 := Keccak256(data)
		if hash1 != hash2 {
			t.Errorf("Keccak256() inconsistent: %v != %v", hash1, hash2)
		}
	})

	t.Run("produce different hashes for different data", func(t *testing.T) {
		data1 := []byte("hello")
		data2 := []byte("world")
		hash1 := Keccak256(data1)
		hash2 := Keccak256(data2)
		if hash1 == hash2 {
			t.Error("Keccak256() produced same hash for different data")
		}
	})
}

func TestKeccak256Str(t *testing.T) {
	t.Run("hash empty string", func(t *testing.T) {
		result := Keccak256Str("")
		expected := "c5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470"
		if result != expected {
			t.Errorf("Keccak256Str() = %v, want %v", result, expected)
		}
	})

	t.Run("hash simple string", func(t *testing.T) {
		result := Keccak256Str("hello")
		if len(result) != 64 {
			t.Errorf("Keccak256Str() length = %v, want 64", len(result))
		}
		matched, _ := regexp.MatchString("^[0-9a-f]{64}$", result)
		if !matched {
			t.Error("Keccak256Str() result is not valid hex")
		}
	})

	t.Run("match Keccak256 with string bytes", func(t *testing.T) {
		s := "test string"
		data := []byte(s)
		if Keccak256Str(s) != Keccak256(data) {
			t.Error("Keccak256Str() doesn't match Keccak256()")
		}
	})
}

func TestHashAliases(t *testing.T) {
	t.Run("Hash is same as Keccak256", func(t *testing.T) {
		data := []byte("test")
		if Hash(data) != Keccak256(data) {
			t.Error("Hash() doesn't match Keccak256()")
		}
	})

	t.Run("HashStr is same as Keccak256Str", func(t *testing.T) {
		if HashStr("test") != Keccak256Str("test") {
			t.Error("HashStr() doesn't match Keccak256Str()")
		}
	})
}

func TestSHA256(t *testing.T) {
	t.Run("hash empty data", func(t *testing.T) {
		data := []byte{}
		result := SHA256(data)
		expected := "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"
		if result != expected {
			t.Errorf("SHA256() = %v, want %v", result, expected)
		}
	})

	t.Run("hash simple data", func(t *testing.T) {
		data := []byte("hello")
		result := SHA256(data)
		if len(result) != 64 {
			t.Errorf("SHA256() length = %v, want 64", len(result))
		}
		matched, _ := regexp.MatchString("^[0-9a-f]{64}$", result)
		if !matched {
			t.Error("SHA256() result is not valid hex")
		}
	})

	t.Run("produce consistent results", func(t *testing.T) {
		data := []byte("test")
		hash1 := SHA256(data)
		hash2 := SHA256(data)
		if hash1 != hash2 {
			t.Errorf("SHA256() inconsistent: %v != %v", hash1, hash2)
		}
	})

	t.Run("produce different hashes for different data", func(t *testing.T) {
		data1 := []byte("hello")
		data2 := []byte("world")
		hash1 := SHA256(data1)
		hash2 := SHA256(data2)
		if hash1 == hash2 {
			t.Error("SHA256() produced same hash for different data")
		}
	})
}

func TestSHA256Str(t *testing.T) {
	t.Run("hash empty string", func(t *testing.T) {
		result := SHA256Str("")
		expected := "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"
		if result != expected {
			t.Errorf("SHA256Str() = %v, want %v", result, expected)
		}
	})

	t.Run("hash simple string", func(t *testing.T) {
		result := SHA256Str("hello")
		if len(result) != 64 {
			t.Errorf("SHA256Str() length = %v, want 64", len(result))
		}
		matched, _ := regexp.MatchString("^[0-9a-f]{64}$", result)
		if !matched {
			t.Error("SHA256Str() result is not valid hex")
		}
	})

	t.Run("match SHA256 with string bytes", func(t *testing.T) {
		s := "test string"
		data := []byte(s)
		if SHA256Str(s) != SHA256(data) {
			t.Error("SHA256Str() doesn't match SHA256()")
		}
	})
}

func TestHashAlgorithmsComparison(t *testing.T) {
	t.Run("Keccak256 and SHA256 produce different hashes", func(t *testing.T) {
		data := []byte("test")
		keccakHash := Keccak256(data)
		sha256Hash := SHA256(data)
		if keccakHash == sha256Hash {
			t.Error("Keccak256() and SHA256() produced same hash")
		}
	})
}

func TestHash(t *testing.T) {
	// Test that Hash function works with different data sizes
	testCases := []struct {
		name string
		data []byte
	}{
		{"empty", []byte{}},
		{"small", []byte("test")},
		{"medium", bytes.Repeat([]byte("a"), 1000)},
		{"large", bytes.Repeat([]byte("b"), 10000)},
	}

	for _, tc := range testCases {
		t.Run(tc.name, func(t *testing.T) {
			result := Hash(tc.data)
			if len(result) != 64 {
				t.Errorf("Hash() length = %v, want 64", len(result))
			}
		})
	}
}
